Reduce unsafe buffer usage in WTF

geoffreygaren opened the pull request:
#### 0b50860edfd4388c393278c559686121406f429e
<pre>
Reduce unsafe buffer usage in WTF
<a href="https://bugs.webkit.org/show_bug.cgi?id=301631">https://bugs.webkit.org/show_bug.cgi?id=301631</a>
<a href="https://rdar.apple.com/163646249">rdar://163646249</a>

Reviewed by Chris Dumez.

* Source/WTF/wtf/Atomics.h:
(WTF::Dependency::consume): Move the unsafe behavior into a helper function in
StdLibExtras to avoid sprinkling unsafe throughout the project and to establish
a self-documenting abstraction that verifies its bounds safety with an ASSERT.

* Source/WTF/wtf/BitSet.h:
(WTF::WordType&gt;::concurrentTestAndSet):
(WTF::WordType&gt;::concurrentTestAndClear): Use operator[] instead of pointer
arithmetic since operator[] is checked.

* Source/WTF/wtf/Brigand.h:
(brigand::detail::count_bools): Use std::array instead of C array, since std::array
is bounds-safe.

* Source/WTF/wtf/BubbleSort.h:
(WTF::bubbleSort): Use span instead of iterators since span is bounds-safe.

* Source/WTF/wtf/CagedPtr.h: Mark our function unsafe so that callers get
labeled unsafe instead of us; we don&apos;t have the information required to make
this behavior safe internally.

* Source/WTF/wtf/CagedUniquePtr.h:
(WTF::CagedUniquePtr::create):
(WTF::CagedUniquePtr::tryCreate): Use span instead of raw pointer because span
is bounds-safe.

* Source/WTF/wtf/CodePtr.h:
(WTF::CodePtr::operator+ const):
(WTF::CodePtr::operator- const):
(WTF::CodePtr::operator+=):
(WTF::CodePtr::operator-=): Mark our function unsafe so that callers get
labeled unsafe instead of us; we don&apos;t have the information required to make
this behavior safe internally.

* Source/WTF/wtf/ConcurrentPtrHashSet.cpp:
(WTF::ConcurrentPtrHashSet::deleteOldTables):
(WTF::ConcurrentPtrHashSet::initialize):
(WTF::ConcurrentPtrHashSet::addSlow):
(WTF::ConcurrentPtrHashSet::containsImplSlow const):
(WTF::ConcurrentPtrHashSet::sizeSlow const):
(WTF::ConcurrentPtrHashSet::resizeIfNecessary):
(WTF::ConcurrentPtrHashSet::Table::create):
(WTF::ConcurrentPtrHashSet::Table::createStub):
(WTF::ConcurrentPtrHashSet::Table::initializeStub): Deleted.
* Source/WTF/wtf/ConcurrentPtrHashSet.h: Use TrailingArray instead of hand-coding
a trailing array. I had to out-of-line the stub table because TrailingArray
must be out of line.
* Source/WTF/wtf/MetaAllocator.cpp:
* Source/WTF/wtf/StdLibExtras.h:
(WTF::addOpaqueZero):

Canonical link: <a href="https://commits.webkit.org/302364@main">https://commits.webkit.org/302364@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aa553257957d613d39645ee883a50f501815a30b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128874 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1131 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39706 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136257 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/80261 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4a173096-5c08-4cac-8395-d4a8e5600708) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1082 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1009 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98114 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/80261 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/8dbc733d-9b9a-48d8-820a-49e324f9db2b) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131821 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/822 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115452 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78728 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c9ed39a8-bc9b-43a1-b653-359381f86b84) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/752 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33563 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79536 "Built successfully") | | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/120884 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109190 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34057 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138718 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/127333 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/947 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/921 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/106663 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1001 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111786 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106474 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/774 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30316 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53404 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20126 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1016 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64326 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/160345 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/859 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40020 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/915 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/951 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/Brigand.h:
```diff
@@ -453,13 +454,10 @@ namespace brigand
-WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
-    constexpr std::size_t count_bools(bool const * const begin, bool const * const end,
-        std::size_t n)
+    constexpr std::size_t count_bools(std::span<const bool> data, std::size_t n)
```
Can't we use `std::ranges::count(data, true)`?

geoffreygaren replied:
Fixed in most recent patch.

cdumez commented on Source/WTF/wtf/CagedPtr.h:
```diff
@@ -78,10 +76,12 @@ class CagedPtr {
     // We need the template here so that the type of U is deduced at usage time rather than class time. U should always be T.
     template<typename U = T>
-    typename std::enable_if<!std::is_same<void, U>::value, T>::type&
+    WTF_UNSAFE_BUFFER_USAGE typename std::enable_if<!std::is_same<void, U>::value, T>::type&
```
Do we really need this in addition to the surrounding `WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN` / `WTF_ALLOW_UNSAFE_BUFFER_USAGE_END`?

geoffreygaren replied:
`WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN` silences the warning here.

`WTF_UNSAFE_BUFFER_USAGE` triggers a warning in caller code.

It's appropriate to trigger a warning in caller code because this access is not bounds-checked, and the callee does not have sufficient information to do a bounds check. To make this bounds-safe, we'll either need to change `CagedPtr` to remembers its bounds, or change the caller to check bounds. `WTF_UNSAFE_BUFFER_USAGE` reminds us to do that work later.

cdumez commented on Source/WTF/wtf/CodePtr.h:
```diff
@@ -207,11 +205,13 @@ class CodePtr : public CodePtrBase {
-    CodePtr operator+(size_t sizeInBytes) const { return CodePtr::fromUntaggedPtr(untaggedPtr<uint8_t*>() + sizeInBytes); }
-    CodePtr operator-(size_t sizeInBytes) const { return CodePtr::fromUntaggedPtr(untaggedPtr<uint8_t*>() - sizeInBytes); }
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
+    WTF_UNSAFE_BUFFER_USAGE CodePtr operator+(size_t sizeInBytes) const { return CodePtr::fromUntaggedPtr(untaggedPtr<uint8_t*>() + sizeInBytes); }
```
ditto.

geoffreygaren replied:
Same reasoning.

cdumez commented on Source/WTF/wtf/ConcurrentPtrHashSet.cpp:
```diff
@@ -185,27 +184,30 @@ bool ConcurrentPtrHashSet::resizeAndAdd(void* ptr)
+
+    // Despite being "full", the stub table allocates a single bucket to allow
+    // loading nullptr from bucket 0.
+    std::unique_ptr<ConcurrentPtrHashSet::Table> result(new (fastMalloc(allocationSize(1))) Table(1));
```
Isn't it UB to allocate the memory with `malloc / fastMalloc` but deallocate with `operator delete` and not `free / fastFree` ?

geoffreygaren replied:
Oops! I believe this was a pre-existing problem. Will fix.

geoffreygaren replied:
Actually, this is fine. The `Table` type already declares `WTF_DEPRECATED_MAKE_STRUCT_FAST_ALLOCATED(Table);`.

geoffreygaren replied:
Placement `new` counts as a `new-expression`, so the behavior of `delete` should not be undefined.

geoffreygaren committed:
Reduce unsafe buffer usage in WTF
https://bugs.webkit.org/show_bug.cgi?id=301631
rdar://163646249

Reviewed by Chris Dumez.

* Source/WTF/wtf/Atomics.h:
(WTF::Dependency::consume): Move the unsafe behavior into a helper function in
StdLibExtras to avoid sprinkling unsafe throughout the project and to establish
a self-documenting abstraction that verifies its bounds safety with an ASSERT.

* Source/WTF/wtf/BitSet.h:
(WTF::WordType>::concurrentTestAndSet):
(WTF::WordType>::concurrentTestAndClear): Use operator[] instead of pointer
arithmetic since operator[] is checked.

* Source/WTF/wtf/Brigand.h:
(brigand::detail::count_bools): Use std::array instead of C array, since std::array
is bounds-safe.

* Source/WTF/wtf/BubbleSort.h:
(WTF::bubbleSort): Use span instead of iterators since span is bounds-safe.

* Source/WTF/wtf/CagedPtr.h: Mark our function unsafe so that callers get
labeled unsafe instead of us; we don't have the information required to make
this behavior safe internally.

* Source/WTF/wtf/CagedUniquePtr.h:
(WTF::CagedUniquePtr::create):
(WTF::CagedUniquePtr::tryCreate): Use span instead of raw pointer because span
is bounds-safe.

* Source/WTF/wtf/CodePtr.h:
(WTF::CodePtr::operator+ const):
(WTF::CodePtr::operator- const):
(WTF::CodePtr::operator+=):
(WTF::CodePtr::operator-=): Mark our function unsafe so that callers get
labeled unsafe instead of us; we don't have the information required to make
this behavior safe internally.

* Source/WTF/wtf/ConcurrentPtrHashSet.cpp:
(WTF::ConcurrentPtrHashSet::deleteOldTables):
(WTF::ConcurrentPtrHashSet::initialize):
(WTF::ConcurrentPtrHashSet::addSlow):
(WTF::ConcurrentPtrHashSet::containsImplSlow const):
(WTF::ConcurrentPtrHashSet::sizeSlow const):
(WTF::ConcurrentPtrHashSet::resizeIfNecessary):
(WTF::ConcurrentPtrHashSet::Table::create):
(WTF::ConcurrentPtrHashSet::Table::createStub):
(WTF::ConcurrentPtrHashSet::Table::initializeStub): Deleted.
* Source/WTF/wtf/ConcurrentPtrHashSet.h: Use TrailingArray instead of hand-coding
a trailing array. I had to out-of-line the stub table because TrailingArray
must be out of line.
* Source/WTF/wtf/MetaAllocator.cpp:
* Source/WTF/wtf/StdLibExtras.h:
(WTF::addOpaqueZero):

Canonical link: https://commits.webkit.org/302364@main
